p2p/background/router_test.go

```diff
@@ -114,6 +123,73 @@ func TestBackgroundRouter_RemovePeer(t *testing.T) {
 	require.Len(t, existingPeerstoreAddrs, 1)
 }
 
+func TestBackgroundRouter_Validation(t *testing.T) {
+	ctx := context.Background()
+	libp2pMockNet := mocknet.New()
+
+	invalidWireFormatData := []byte("test message")
+	invalidPocketEnvelope := &anypb.Any{
+		TypeUrl: "/test",
+		Value:   invalidWireFormatData,
+	}
+	invalidPocketEnvelopeBz, err := proto.Marshal(invalidPocketEnvelope)
+	require.NoError(t, err)
+
+	invalidMessages := [][]byte{
+		invalidWireFormatData,
+		invalidPocketEnvelopeBz,
+	}
+
+	receivedChan := make(chan struct{})
+
+	receiverPrivKey, receiverPeer := newTestPeer(t)
+	receiverHost := newTestHost(t, libp2pMockNet, receiverPrivKey)
+	receiverRouter := newRouterWithSelfPeerAndHost(t, receiverPeer, receiverHost, func(data []byte) error {
+		receivedChan <- struct{}{}
+		return nil
+	})
+
+	t.Cleanup(func() {
+		err := receiverRouter.Close()
+		require.NoError(t, err)
+	})
+
+	senderPrivKey, _ := newTestPeer(t)
+	senderHost := newTestHost(t, libp2pMockNet, senderPrivKey)
+	gossipPubsub, err := pubsub.NewGossipSub(ctx, senderHost)
+	require.NoError(t, err)
+
+	err = libp2pMockNet.LinkAll()
+	require.NoError(t, err)
+
+	receiverAddrInfo, err := utils.Libp2pAddrInfoFromPeer(receiverPeer)
+	require.NoError(t, err)
+
+	err = senderHost.Connect(ctx, receiverAddrInfo)
+	require.NoError(t, err)
+
+	topic, err := gossipPubsub.Join(protocol.BackgroundTopicStr)
+	require.NoError(t, err)
+
+	for _, invalidMessageBz := range invalidMessages {
+		err = topic.Publish(ctx, invalidMessageBz)
+		require.NoError(t, err)
+	}
+
+	select {
+	case <-time.After(time.Second * 2):
+		// TECHDEBT: find a better way to prove that pre-propagation validation
+		// works as expected. Ideally, we should be able to distinguish which
```

Comment:
Wdym by `Ideally, we should be able to distinguish which invalid message was received in the event of failure.`

Can't we filter based on the message type/content?

Reply:
Thanks for pressing this, there has been a lot of shifting ground around these parts so it's good that you made me revisit! :pray: 

The issue here currently is that we're passing a method of `BackgroundRouter` to [`Gossipsub#RegisterTopicValidator()`](https://pkg.go.dev/github.com/libp2p/go-libp2p-pubsub#section-readme) which doesn't leave any room for the test to hook in. After looking at the docs again though, I'm realizing that I can use `#UnregisterTopicValidator()` in the test and subsequently re-register a wrapper function which the test can hook into. :+1:

Reply:
I re-wrote the test, apologize the the additional review effort. I think it's much higher quality now though - I hope you agree. :sweat_smile: